node/network/statement-distribution/src/vstaging/requests.rs

```diff
@@ -213,7 +213,24 @@ impl RequestManager {
 		}
 	}
 
-	// TODO [now]: removal based on relay-parent.
+	/// Remove based on relay-parent.
+	pub fn remove_by_relay_parent(&mut self, relay_parent: Hash) {
+		// Remove from `by_priority` and `requests`.
+		self.by_priority.retain(|(_priority, id)| {
+			let retain = relay_parent != id.relay_parent;
+			if !retain {
+				self.requests.remove(id);
+			}
+			retain
+		});
+		// Remove from `unique_identifiers`.
+		for (_, candidate_identifier_set) in self.unique_identifiers.iter_mut() {
+			candidate_identifier_set.retain(|id| relay_parent != id.relay_parent);
+		}
+		// TODO: is this necessary?
+		// If any candidates don't have unique identifiers, remove them.
+		self.unique_identifiers.retain(|_, set| !set.is_empty());
```

Comment:
Wasn't totally sure about this.

Reply:
Seems correct, since we would keep around empty `unique_identifiers` sets over time and gradually leak memory.

Reply:
I think the implementation could be made more efficient with some code like this:

1. In the first step, when pruning `by_priority`, construct a `Vec<CandidateHash>` or `HashSet` of candidates to inspect in `unique_identifiers`.
2. Loop over this set, invoking `self.unique_identifiers.entry(candidate_hash)`, and then in a single branch on `Entry::Occupied`
    a. remove all entries with the `relay_parent` from the candidate identifier set
    b. if the set is empty, invoke `entry.remove()`
    c. we can ignore `Entry::Vacant`. I can explain why we can expect to encounter it but only if nodes are misbehaving and we don't use a deduplicating collection, and there are no issues from ignoring it